src/ctap/data_formats.rs

```diff
@@ -445,66 +445,91 @@ pub struct PublicKeyCredentialSource {
     pub user_handle: Vec<u8>, // not optional, but nullable
     pub other_ui: Option<String>,
     pub cred_random: Option<Vec<u8>>,
+
+    /// Contains the unknown fields when parsing a CBOR value.
+    ///
+    /// Those fields could either be deleted fields from older versions (they should have reserved
+    /// tags) or fields from newer versions (the tags should not be reserved). If this is empty,
+    /// then the parsed credential is probably from the same version (but not necessarily).
+    pub unknown_fields: BTreeMap<cbor::KeyType, cbor::Value>,
+}
+
+// We simulate protocol buffers in CBOR with maps. Each field of a message is associated with a
+// unique tag, implemented with a CBOR unsigned key.
```

Comment:
There's nothing protocol-buffer-specific in this. FIDO2 already encodes many structs as maps, with string or integer keys, for example:
https://github.com/google/OpenSK/blob/8bbf42623f57f09c2c17ed5acd26f77eb49d6545/src/ctap/data_formats.rs#L1127-L1131

Protocol buffers refers to a [specific encoding of tag-value pairs](https://developers.google.com/protocol-buffers/docs/encoding), which this change doesn't implement at all (the CBOR encoding is used instead). So referring to "protocol buffers" here is misleading/confusing.

I'd simply mention "Encode credentials as a map, for versioning" or something like that.

Reply:
It's mostly the ideas:
- Do not reuse an tag that existed in any past versions.
- Provide access to possibly older/newer fields (to detect possible issues).

Anyone familiar with proto-buf would recognize those patterns. If we believe those are common knowledge, I can remove the mention to proto-buf and just keep the ideas.

Reply:
I don't think this pattern (tag + value) is specific to protobufs (https://en.wikipedia.org/wiki/Type-length-value). Sure, protobufs use it (so "anyone familiar with proto-buf would recognize those patterns"), but so do many encoding formats, and I don't think knowledge of protobufs should be required to understand what's being done here.

Just saying that we replace an array by a mapping from (versioned) tags to values would be enough.

Reply:
It's not only the tag+value, it's alse the [reserved fields](https://developers.google.com/protocol-buffers/docs/proto#reserved) and [unknown fields](https://developers.google.com/protocol-buffers/docs/proto#updating) support. All of those are supported by default in proto. Here we have to encode the tags ourself. There is no description of the message.

Reply:
Reserved/unknown fields are also not specific to protobufs. Plenty of formats have those. For example PNG images (https://en.wikipedia.org/wiki/Portable_Network_Graphics#File_format) have a concept of "critical" (i.e. mandatory) and "ancillary" (i.e. vendor-specific, potentially unknown to a specific decoder) fields.

Again, protobufs mostly refers to a specific encoding on the wire, which we don't use here at all, since we use the CBOR encoding on the wire.

> Here we have to encode the tags ourself. There is no description of the message.

So another difference with protos is that we don't even have such a description of the message, right?

I think mentioning protobufs at all is just confusing - we can just explicitly mention the concepts of reserved/unknown fields, or simply the concept of "versioning".

Reply:
> So another difference with protos is that we don't even have such a description of the message, right?

This is exactly why it's useful to mention it, because the concept of description is missing. And it's an important concept to understand to be able to understand the code. But I just removed mentions to protobufs since it looks like it's more problematic than anything else.